File: members/U02CDR60PFS/programmers/방의개수.cpp
#include <string>
#include <vector>
#include <iostream>
#include <map>
using namespace std;
int dx[8] = {0, 1, 1, 1, 0, -1, -1, -1};
int dy[8] = {-1, -1, 0, 1, 1, 1, 0, -1};
int solution(vector<int> arrows)
{
    int answer = 0;
    map<pair<int, int>, bool> node;
    map<pair<pair<int, int>, pair<int, int>>, bool> edge;

    int x = 0, y = 0;
    node[{x, y}] = true;
    for (int arrow : arrows)
    {
        for (int i = 0; i < 2; i++)
        {
            int nx = x + dx[arrow];
            int ny = y + dy[arrow];

            if (node[{nx, ny}] == true && edge[{{x, y}, {nx, ny}}] == false)
            {
                answer++;
            }
            node[{nx, ny}] = true;
            edge[{{x, y}, {nx, ny}}] = true;
            edge[{{nx, ny}, {x, y}}] = true;
            x = nx;
            y = ny;
        }
    }

    return answer;
}

File: members/U02BWCHJA6Q/Programers/practice/방의_개수.js
function solution(arrows) {
  let answer = 0;
  const direct = [
    [-1, 0],
    [-1, 1],
    [0, 1],
    [1, 1],
    [1, 0],
    [1, -1],
    [0, -1],
    [-1, -1],
  ];
  const visited = new Set();
  const lines = new Set();
  let y = 0;
  let x = 0;
  visited.add('0-0');

  arrows.forEach((arrow) => {
    const nextY = y + direct[arrow][0];
    const nextX = x + direct[arrow][1];

    const key = `${nextY}-${nextX}`;
    const line = `${y}-${x}-${nextY}-${nextX}`;

    if (lines.has(line)) {
      y = nextY;
      x = nextX;

      return;
    }

    if (visited.has(key)) {
      answer++;
    }

    if (arrow === 1) {
      const reverseY = y - 1;
      const reverseX = x;
      const reverseNextY = y;
      const reverseNextX = x + 1;

      const reverseLine = `${reverseY}-${reverseX}-${reverseNextY}-${reverseNextX}`;

      if (lines.has(reverseLine)) {
        answer++;
      }
    } else if (arrow === 3) {
      const reverseY = y + 1;
      const reverseX = x;
      const reverseNextY = y;
      const reverseNextX = x + 1;

      const reverseLine = `${reverseY}-${reverseX}-${reverseNextY}-${reverseNextX}`;

      if (lines.has(reverseLine)) {
        answer++;
      }
    } else if (arrow === 5) {
      const reverseY = y;
      const reverseX = x - 1;
      const reverseNextY = y + 1;
      const reverseNextX = x;

      const reverseLine = `${reverseY}-${reverseX}-${reverseNextY}-${reverseNextX}`;

      if (lines.has(reverseLine)) {
        answer++;
      }
    } else if (arrow === 7) {
      const reverseY = y;
      const reverseX = x - 1;
      const reverseNextY = y - 1;
      const reverseNextX = x;

      const reverseLine = `${reverseY}-${reverseX}-${reverseNextY}-${reverseNextX}`;

      if (lines.has(reverseLine)) {
        answer++;
      }
    }

    const reverseLine = `${nextY}-${nextX}-${y}-${x}`;

    visited.add(key);
    lines.add(line);
    lines.add(reverseLine);

    y = nextY;
    x = nextX;
  });

  return answer;
}
